docs/proposals/11-apiserver-network-proxy.md

```diff
@@ -0,0 +1,134 @@
+
+The initial motivation for the apiserver-network-proxy project is to get rid of provider-specific implementations that reside in the API-server (e.g., SSH), but it turns out that
+it has other interesting use-cases such as data-plane connection decoupling, which is the main use-case for this proposal.
+
+Starting with **Kubernetes 1.18** it's possible to make use of an `--egress-selector-config-file` flag, this helps point the API-server to traffic hook points based on traffic direction. For example, in the config below the API server would have to forward all cluster related traffic (e.g., logs, port-forward, exec, ...etc.) to the **proxy-server** which then knows how to forward traffic to the shoot. For the rest of the traffic, e.g. API server to ETCD or other control-plane components `direct` is used which means legacy routing method, i.e., by-pass the proxy.
+```yaml
+  egress-selector-configuration.yaml: |-
+    apiVersion: apiserver.k8s.io/v1alpha1
+    kind: EgressSelectorConfiguration
+    egressSelections:
+    - name: cluster
+      connection:
+        proxyProtocol: httpConnect
+        transport:
+          tcp:
+            url: https://proxy-server:8131
+    - name: master
+      connection:
+        proxyProtocol: direct
+    - name: etcd
+      connection:
+        proxyProtocol: direct
+```
+
+## Challenges
+
+
+### Prometheus to Shoot connectivity
+One challenge remains to completely eliminate the need for a VPN connection. In today's Gardener setup, each control-plane has a Prometheus instance that directly scrapes cluster components such as CoreDNS, Kubelets, cadvisor, etc. This works because in addition to the VPN side car attached to the API server pod, we have another one attached to prometheus which knows how to forward traffic to these endpoints. Once the VPN is eliminated, it is required to find other means to forward traffic to these components.
+
+#### Possible Solutions
+
+There are currently two ways to solve this problem:
+
+- Attach a port-forwarder side-car to prometheus.
+- Utilize the proxy subresource on the API server.
+
+#### Port-forwarder Sidecar
+
+With this solution each prometheus instance would have a side-car that has the kubeconfig of the shoot cluster, and which estalishes a port-forward connection to the endpoints residing in the shoot.
+
+There are a couple of problems with this approach:
+
+- the port-forward connection is not reliable.
+- the connection would break if the API server instance dies.
+- requires an additional component.
+
+#### Proxy sub-resource
+
+Kubernetes supports proxying requests to nodes, services, and pod endpoints in the shoot cluster. This proxy connection can be utilized for scraping the necessary endpoints in the shoot.
+
+This approach requires less-compenents and is more reliable than the port-forward solution, however, it relies on having the API server supporting proxied connection for the required endpoints.
```

Comment:
It also comes with additional load on the API server (which we don't have today), right? Is this negligible?
Which solution do you prefer?

Reply:
it add additional load to the API-server yes but I don't think its significat, also the scrape interval from prometheus has been halfed  so it shouldn't be much. This solution is the most achievable, using port-forward soltuion is also putting load on the API but would require knowledge of all services in the shoot cluster that we need to port-forward to, so I prefer the proxy solution in this case.